src/iosrv/sched.c

```diff
@@ -26,73 +26,953 @@
+		     void (*func)(void *), void *arg)
+{
+	ABT_pool	abt_pool;
+	int		rc;
+
+	D_ASSERT(attr && func && arg);
+	switch (attr->sra_type) {
+	case SCHED_REQ_IO:
+		abt_pool = dx->dx_pools[DSS_POOL_IO];
+		break;
+	case SCHED_REQ_GC:
+		abt_pool = dx->dx_pools[DSS_POOL_GC];
+		break;
+	case SCHED_REQ_MIGRATE:
+		abt_pool = dx->dx_pools[DSS_POOL_REBUILD];
+		break;
+	default:
+		D_ASSERTF(0, "Invalid req type: %u\n", attr->sra_type);
+		break;
+	}
+
+	rc = ABT_thread_create(abt_pool, func, arg, ABT_THREAD_ATTR_NULL, NULL);
+
+	return dss_abterr2der(rc);
+}
+
+static int
+req_kickoff(struct dss_xstream *dx, struct sched_request *req)
+{
+	struct sched_info	*info = &dx->dx_sched_info;
+	struct sched_pool_info	*spi = req->sr_pool_info;
+	struct sched_req_info	*sri;
+	int			 rc;
+
+	if (req->sr_ult != ABT_THREAD_NULL) {
+		rc = ABT_thread_resume(req->sr_ult);
+		rc = dss_abterr2der(rc);
+	} else {
+		rc = req_kickoff_internal(dx, &req->sr_attr, req->sr_func,
+					  req->sr_arg);
+	}
+
+	D_ASSERT(spi != NULL);
+	D_ASSERT(req->sr_attr.sra_type < SCHED_REQ_MAX);
+	sri = &spi->spi_req_array[req->sr_attr.sra_type];
+
+	D_ASSERT(sri->sri_req_cnt > 0);
+	sri->sri_req_cnt--;
+	D_ASSERT(info->si_req_cnt > 0);
+	info->si_req_cnt--;
+
+	d_list_del_init(&req->sr_link);
+	req_put(dx, req);
+
+	return rc;
+}
+
+#define SCHED_SPACE_AGE_MAX	5000	/* 5000 msecs */
+
+static int
+check_space_pressure(struct dss_xstream *dx, struct sched_pool_info *spi,
+		     bool purge)
+{
+	struct sched_info	*info = &dx->dx_sched_info;
+	struct vos_pool_space	 vps = { 0 };
+	uint64_t		 scm_left, nvme_left;
+	int			 rc;
+
+	D_ASSERT(spi->spi_space_ts <= info->si_cur_ts);
+	/* Use cached space presure info */
+	if ((spi->spi_space_ts + SCHED_SPACE_AGE_MAX) > info->si_cur_ts)
+		goto out;
+
+	rc = vos_pool_query_space(spi->spi_pool_id, &vps);
+	if (rc == -DER_NONEXIST) {	/* vos pool is destroyed */
+		D_CDEBUG(purge, DB_TRACE, DLOG_ERR,
+			 "XS(%d): vos pool:"DF_UUID" is destroyed.\n",
+			 dx->dx_xs_id, DP_UUID(spi->spi_pool_id));
+		if (purge)
+			add_purge_list(dx, spi);
+		goto out;
+	} else if (rc) {
+		D_ERROR("XS(%d): query pool:"DF_UUID" space failed. "DF_RC"\n",
+			dx->dx_xs_id, DP_UUID(spi->spi_pool_id), DP_RC(rc));
+		goto out;
+	}
+	spi->spi_space_ts = info->si_cur_ts;
+
+	D_ASSERT(SCM_SYS(&vps) < SCM_TOTAL(&vps));
+	/* NVME_TOTAL and NVME_SYS could be both zero */
+	D_ASSERT(NVME_SYS(&vps) <= NVME_TOTAL(&vps));
+
+	if (SCM_FREE(&vps) > SCM_SYS(&vps))
+		scm_left = SCM_FREE(&vps) - SCM_SYS(&vps);
+	else
+		scm_left = 0;
+
+	if (NVME_FREE(&vps) > NVME_SYS(&vps))
+		nvme_left = NVME_FREE(&vps) - NVME_SYS(&vps);
+	else
+		nvme_left = 0;
+
+	if (scm_left < (SCM_TOTAL(&vps) * 1 / 10) ||
+	    nvme_left < (NVME_TOTAL(&vps) * 1 / 10))
+		spi->spi_space_pressure = 2;	/* severe space pressure */
+	else if (scm_left < (SCM_TOTAL(&vps) * 3 / 10) ||
+		 nvme_left < (NVME_TOTAL(&vps) * 3 / 10))
+		spi->spi_space_pressure = 1;	/* light space pressure */
```

Comment:
maybe can "#define DAOS_SPACE_PRESSURE_LIGHT 1" or sth else and use that also in process_pool_cb().

Reply:
Ok, sure.